### PR TITLE
Feature/main files

### PR DESCRIPTION
**Pull request by bt-s**
Have a look at whether you approve of the changes that I have suggested. Also, if you think that the docstrings are incorrect, please leave a note why you think so and request for changes or change it yourself.

NOTE: all code regarding tiny ImageNet has been removed from this branch, for clarity's sake.
If we want to use this dataset, it can be added again later, and the relevant code can be found in the Git history.

### CODE REVIEW

File: RCNN_tf.py [hunk 1]
```diff
@@ -68,46 +93,81 @@ def rcl(
...
+
             x += 1
+
             return x, recurrent_states, output
 
         results = tf.while_loop(
-            lambda x, recurrent_states, output: x < depth,
-            loop_body,
-            [0, cell_states, output],
+            cond=termination_cond(),
```
Comment (leoTiez) [inline RCNN_tf.py]: I don't think this works. `x`, `recurrent_states`, and `output` should be passed as parameter.
I will change that

File: RCNN_tf.py [hunk 2]
```diff
@@ -21,25 +26,39 @@
...
+    """Defines a recurrent convolutional layer
+
+    Args:
+        input_data   (tf.Tensor): 4D data tensor
+        num_input_chans    (int): the number of input channels
+        num_filter         (int): the number of filters
+        filter_shape      (list): the shape of the 2D filter
+        num_of_data  (tf.Tensor): holds the batch size
+        processing_unit    (str): specifies whether a GPU or CPU is used
+        depth              (int): depth of the ???
```
Comment (leoTiez) [inline RCNN_tf.py]: I will add the description here

File: RCNN_tf.py [hunk 3]
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
...
+    """Defines a convolutional layer
+
+    Args:
+        input_data   (tf.Tensor): 4D data tensor
+        num_input_chans    (int): the number of input channels
+        num_filter         (int): the number of filters
+        filter_shape      (list): the shape of the 2D filter
+        stddev           (float): standard deviation
+        stride           (tuple): ???
+        padding            (str): ???
```
Comment (leoTiez) [inline RCNN_tf.py]: I can add the docstrings. For further information about stride and padding:
https://www.quora.com/What-are-the-roles-of-stride-and-padding-in-a-convolutional-neural-network

File: RCNN_tf.py [hunk 4]
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
...
-        padding='same'
-):
+def pooling_layer(input_data, pool_shape, stride=(1, 1), padding='same'):
+    """Defines a pooling layer
+
+    Args:
+        input_data (tf.Tensor): 4D data tensor
+        pool_shape      (list): the shape of the 2D filter
+        stride         (tuple): ???
+        padding          (str): ???
```
Comment (leoTiez) [inline RCNN_tf.py]: Same here

File: RCNN_tf.py [hunk 5]
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
...
-            pooling_stride_shape=[2, 2]
-    ):
+            shuf_buf_size=10000,
+            conv_filter_shape=[5, 5],
+            rconv_filter_shape=[3, 3],
+            pool_shape=[3, 3],
+            pool_stride_shape=[2, 2]
+        ):
+
+        # Input and output shape placeholders
```
Comment (leoTiez) [inline RCNN_tf.py]: That's not true. These are placeholders for the input and output data. However, the known dimensions should be set beforehand. This is why the shape is explicitly set. Since we want to be flexible with the number of inputs we can pass `None`

File: RCNN_tf.py [hunk 6]
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
...
+            self.output_placeholder)
+            ).shuffle(buffer_size=shuf_buf_size
+            ).repeat(
+            ).batch(batch_size=self.num_data_placeholder)
+
         test_data_set = tf.data.Dataset.from_tensor_slices((
             self.input_placeholder,
-            self.output_placeholder,
-        )).batch(batch_size=self.num_data_placeholder)
+            self.output_placeholder,)
```
Comment (leoTiez) [inline RCNN_tf.py]: I will remove the comma here

File: RCNN_tf.py [hunk 7]
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
...
 
 
 class RCNN:
+    """The Recurrent Convolutional Neural Network Classifier
+
+    Attributes:
+        input_shape        (list): specification of the image input shape
+        output_shape       (list): specification of the class output shape
+        processing_unit     (str): specification of the PU to be used
+        learning_rate     (float): learning rate
```
Comment (leoTiez) [inline RCNN_tf.py:319]: Aren't the docstrings for the constructor normally within the constructor?

File: RCNN_tf.py [hunk 8]
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
...
+                pool_shape=pool_shape,
+                stride=pool_stride_shape
             )
 
-            # Recurrent convolutional layers
+            # First recurrent convolutional layer
             rcl_layer_1 = rcl(
                 input_data=pooling_1,
-                num_input_channels=num_filter,
+                num_input_chans=num_filter,
```
Comment (leoTiez) [inline RCNN_tf.py:400]: Fine for now, but in future I would like to avoid renaming variables, if there is no reason for that. It is totally appropriate when they are deprecated, misleading, or not descriptive.
This, however, adds points of failure and blows up the changes on github which are needed to be reviewed.

File: RCNN_tf.py [hunk 9]
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
...
                 num_of_data=self.num_data_placeholder,
-                device_name=self.device_name,
-                filter_shape=rec_conv_layer_filter_shape,
```
Comment (leoTiez) [inline RCNN_tf.py:266]: Same with this one

File: RCNN_tf.py [hunk 10]
```diff
@@ -396,93 +548,168 @@ def train(
...
+
                     avg_cost += cost_ / total_batch
                     train_writer.add_summary(accuracies)
 
                 sess.run(self.test_init_op, feed_dict={
-                    self.input_placeholder: val_data_features,
+                    self.input_placeholder: val_data_feats,
                     self.output_placeholder: val_data_labels,
-                    self.num_data_placeholder: val_data_features.shape[0]
+                    self.num_data_placeholder: val_data_feats.shape[0]
```
Comment (leoTiez) [inline RCNN_tf.py:578]: Again: Fine for now, but this can be very dangerous and break things.

File: RCNN_tf.py [hunk 11]
```diff
@@ -396,93 +548,168 @@ def train(
...
+
+        test_data_, test_labels_ = data_loader(
+            "cifar100",
+            "test",
+            dtype=PRECISION_NP
+        )
+
+    else:
+        raise Exception("dataset has to be one of 'mnist', 'cifar10', "
+                        "'cifar100' or 'tinynet'.")
```
Comment (leoTiez) [inline RCNN_tf.py]: I will remove the tinynet

File: RCNN_tf.py [hunk 12]
```diff
@@ -396,93 +548,168 @@ def train(
...
-        test_labels_,
-        epochs=epochs_
+        train_data_feats=training_data_,
+        train_data_labels=training_labels_,
+        val_data_feats=test_data_,
+        val_data_labels=test_labels_,
+        #batch_size=100,
+        #epochs=epochs_,
+        #create_graph=True,
+        #print_vars=True
```
Comment (leoTiez) [inline RCNN_tf.py:712]: Is there a reason why they are commented out?

File: RCNN_tf.py [hunk 13]
```diff
@@ -10,9 +10,14 @@
 __author__ = "Adrian Chmielewski-Anders, Leo Zeitler & Bas Straathof"
 
 import sys
+import os
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2' # suppress warnings
+
 import tensorflow as tf
+tf.logging.set_verbosity(tf.logging.ERROR) # suppress warnings
```
Comment (leoTiez) [inline RCNN_tf.py:17]: Are both lines
`os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'``and
`tf.logging.set_verbosity(tf.logging.ERROR) `
really necessary to suppress the warnings or is it more duplicating

File: RCNN_tf.py [hunk 14]
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
...
 
         # Create data set objects
         training_data_set = tf.data.Dataset.from_tensor_slices((
             self.input_placeholder,
-            self.output_placeholder
-        )).shuffle(buffer_size=shuffle_buffer_size).repeat().batch(batch_size=self.num_data_placeholder)
+            self.output_placeholder)
+            ).shuffle(buffer_size=shuf_buf_size
+            ).repeat(
+            ).batch(batch_size=self.num_data_placeholder)
```
Comment (leoTiez) [inline RCNN_tf.py]: This line splitting is very hard to read I find

File: RCNN_tf.py (hunk 1 above)
Comment (bt-s) [reply RCNN_tf.py]: I tested it and it seems to work just fine -- in my opinion it makes the code a bit easier to understand. The function termination_cond returns x, recurrent_states and output.

File: RCNN_tf.py (hunk 3 above)
Comment (bt-s) [reply RCNN_tf.py]: Ah, yeah: I know what these mean haha.
I just forgot to come back to it, my bad.

File: RCNN_tf.py (hunk 5 above)
Comment (bt-s) [reply RCNN_tf.py]: Good!

File: RCNN_tf.py (hunk 10 above)
Comment (bt-s) [reply RCNN_tf.py:578]: I did a thorough test, so don't worry.

File: RCNN_tf.py (hunk 11 above)
Comment (bt-s) [reply RCNN_tf.py]: Well spotted!

File: RCNN_tf.py (hunk 12 above)
Comment (bt-s) [reply RCNN_tf.py:712]: Yes: I just wanted to signal that these are the standard input parameters.
They could be changed when we are tweaking the model.

File: RCNN_tf.py (hunk 13 above)
Comment (bt-s) [reply RCNN_tf.py:17]: They are both necessary, since one suppresses warnings and the other one suppress the "I, Tensorflow did .... bla bla bla" output.

File: RCNN_tf.py (hunk 14 above)
Comment (bt-s) [reply RCNN_tf.py]: Yeah, I agree: change it if you have a better formatting in mind.

In Javascript I would write something like:

```
self.output_placeholder))
     .shuffle(buffer_size=shuf_buf_size)
     .repeat()
     .batch(batch_size=self.num_data_placeholder)
```

Unfortunately Python cannot parse that....

**Comment by leoTiez** [comment]
I will add `Pillow` to the requirements.txt. Please add every library to the requirements.txt, if it is not a standard library

**Comment by leoTiez** [comment]
But I see right now you do not even use it. Alright, then I will remove these lines

**Comment by leoTiez** [comment]
All the issues have been discussed. Ready for merge

**Review by leoTiez — APPROVED** [review]
